File: vercors/src/main/java/vct/col/ast/ASTDeclaration.java
package vct.col.ast;

import vct.util.ClassName;

public abstract class ASTDeclaration extends ASTNode {

  public abstract ClassName getDeclName();

  /** contains the name of the class/module/package. */
  public final String name;

  protected ASTDeclaration(String name){
    this.name=name;
  }
  
  /** contains the root of the source forest. */
  protected ProgramUnit root;
  /** contains the package name */
  protected ClassName package_name;
  
  public boolean packageDefined(){
    return package_name!=null;
  }
  public void attach(ProgramUnit root,ClassName package_name){
    this.root=root;
    this.package_name=package_name;
  }

}

File: vercors/src/main/java/vct/col/ast/stmt/decl/ASTDeclaration.scala
package vct.col.ast.stmt.decl

import vct.col.ast.generic.ASTNode
import vct.util.ClassName;

/**
 * Abstract AST node that represents a declaration.
 * 
 * @param name Contains the name of the class/module/package.
 */
abstract class ASTDeclaration(val name:String) extends ASTNode {
  /** Contains the root of the source forest. */
  protected var root:ProgramUnit = null
  
  /** Contains the package name. */
  protected var packageName:ClassName = null
  
  def getDeclName : ClassName
  def packageDefined = packageName != null
  
  def attach(root:ProgramUnit, packageName:ClassName) = {
    this.packageName = packageName
    this.root = root
  }
}
